### PR TITLE
Feature/database

### PR DESCRIPTION
**Pull request by Boukhari-Mohamed**

### CODE REVIEW

File: app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/Database.java [hunk 1]
```diff
@@ -1,9 +1,48 @@
...
 public interface Database {
 
+
+    /**
+     * Fetches the requested data
+     * @param collectionPath collection
+     * @param documentPath document id
+     * @return task that fails if the database is unreachable
+     */
+    Task<DocumentSnapshot> fetch(String collectionPath, String documentPath);
```
Comment (khalilacheche) [inline app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/Database.java:14]: Using DocumentSnpashot as the return type of the Task might be an abstraction leak, as these types are specific to firebase. I dont think its worth it for now to switch to a more abstract implementation, though

File: app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/dummy/DummyDatabase.java [hunk 2]
```diff
@@ -61,4 +63,28 @@ public void reset() {
...
+
+    @Override
+    public Task<String> add(String collectionPath, Object data) {
+        return null;
+    }
+
+    @Override
+    public Task<CollectionSnapshot> fetchAll(String collectionPath) {
+        return null;
+    }
```
Comment (khalilacheche) [inline app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/dummy/DummyDatabase.java:89]: these methods shouldn't be retuning null, it should be at least returning dummy tasks

**Review by khalilacheche — APPROVED** [review]
The implementation is overall really good, the classes are designed well and fit the chosen architecture. The main issue is about tests.

File: app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/ui/activities/ChatRoomActivity.java [hunk 3]
```diff
@@ -35,8 +36,8 @@ public class ChatRoomActivity extends AppCompatActivity{
     @Override
     protected void onCreate(Bundle savedInstanceState) {
 
-        DummyDatabase dep = DummyDatabase.getInstance();
-        messageList = new ArrayList<>(dep.fetchMessages());
+        Database dep = DatabaseFactory.getDependency();
```
[the remark this answers is not in the document]
Comment (Boukhari-Mohamed) [reply app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/ui/activities/ChatRoomActivity.java:39]: Yes, Some refractoring to incorporate the new changes should be done for a lot of activities not just this one. But those lines could be removed, they were there just for demo purposes.

File: app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/dummy/DummyDatabase.java (hunk 2 above)
Comment (Boukhari-Mohamed) [reply app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/dummy/DummyDatabase.java:89]: The dummy db will be implemented correctly later, after discussing with ellior and the rest of the team.

File: app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/firebase/FirebaseDatabase.java [hunk 4]
```diff
@@ -55,15 +47,15 @@ public Task<Void> delete(String collectionPath, String documentPath) {
...
     @Override
-    public Task<String> add(String collectionPath, Map<String, Object> data) {
+    public Task<String> add(String collectionPath, Object data) {
         return database.collection(collectionPath)
                 .add(data).onSuccessTask((documentReference ->
                         Tasks.forResult(documentReference.getId())
                 ));
     }
 
     @Override
```
Comment (badismach) [inline app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/firebase/FirebaseDatabase.java:57]: Are you planning on implementing the fetchAll function? Maybe add a comment to specify why it's empty for now.

**Review by badismach — APPROVED** [review]
Overall good implementation. As with most of our code, comments are lacking. Please comment for better readability.

File: app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/firebase/FirebaseDatabase.java [hunk 5]
```diff
@@ -55,15 +47,15 @@ public Task<Void> delete(String collectionPath, String documentPath) {
...
-    public Task<DBDataCollection> fetchAll(String collectionPath) {
+    public Task<CollectionSnapshot> fetchAll(String collectionPath) {
         return null;
     }
 }
```
Comment (badismach) [inline app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/firebase/FirebaseDatabase.java:61]: Are you planning on implementing the fetchAll function? Maybe add a comment to specify why it's empty for now.

File: app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/models/User.java [hunk 6]
```diff
@@ -43,6 +44,9 @@ public User(String id, String email, String firstName, String lastName) {
         setLastName(lastName);
     }
 
+    public User(){
+    }
```
Comment (raedbm) [inline app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/models/User.java:48]: why is this empty?

**Review by raedbm — APPROVED** [review]
Overall the database implementation is going well. I don't think I can have remarks on the implementation in itself as you are way more informed about the implementation of the database than me, However if you could look into making tests as only one test for the dependency was done.

**Pull request merged**